src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -217,20 +217,15 @@ let PostParseModuleSpec (_i, defaultNamespace, isLastCompiland, fileName, intf)
         SynModuleOrNamespaceSig(lid, isRecursive, kind, decls, xmlDoc, attributes, None, range, trivia)
 
 let GetScopedPragmasForHashDirective hd (langVersion: LanguageVersion) =
-    let supportsNonStringArguments =
-        langVersion.SupportsFeature(LanguageFeature.ParsedHashDirectiveArgumentNonQuotes)
-
     [
         match hd with
-        | ParsedHashDirective("nowarn", numbers, m) ->
-            for s in numbers do
+        | ParsedHashDirective("nowarn", args, m) ->
+            for s in args do
                 let warningNumber =
-                    match supportsNonStringArguments, s with
-                    | _, ParsedHashDirectiveArgument.SourceIdentifier _ -> None
-                    | true, ParsedHashDirectiveArgument.LongIdent _ -> None
-                    | true, ParsedHashDirectiveArgument.Int32(n, _) -> GetWarningNumber(m, string n, true)
-                    | true, ParsedHashDirectiveArgument.Ident(s, _) -> GetWarningNumber(m, s.idText, true)
-                    | _, ParsedHashDirectiveArgument.String(s, _, _) -> GetWarningNumber(m, s, true)
+                    match s with
+                    | ParsedHashDirectiveArgument.Int32(n, m) -> GetWarningNumber(m, string n, langVersion, false)
+                    | ParsedHashDirectiveArgument.Ident(s, m) -> GetWarningNumber(m, s.idText, langVersion, false)
+                    | ParsedHashDirectiveArgument.String(s, _, m) -> GetWarningNumber(m, $"\"{s}\"", langVersion, false)
```

Comment:
I know why it's there, but TBH it feels a little like scratching my left ear with my right foot:

The lexer does the work to parse a string and gets rid of the quotes.
This line of code adds the quotes back.
The code inside GetWarningNumber uses regex to get rid of the quotes again.


I think there should be  way of 2 of these steps cancelling each other out, and not doing unneccessary string concats + parsing.

Reply:
Yes, I do understand that this looks strange.
The new `GetWarningNumber` comes from my ongoing work on "scoped nowarn". Since there we want fine-grained scopes, #nowarn and #warnon must be allowed anywhere (also inside sub-modules and expressions). This means that we must process these directives in the lexer rather than the parser. If we do that, all the strange code will disappear.

But probably this means I should not propose the new `GetWarningNumber` now, but only with the "scoped nowarn" PR.
Please advice.

So, what will remain are the first two items mentioned in #17870.
Also for these two items, however, I am not sure, given the other feedback that I got, if they are welcome.
So, also here I need advice.
I have no problems scratching the whole PR and bringing parts of it back with "scoped nowarn".

Reply:
I think there is a way to let this in now already.
What if the GetWarningNumber did not work on stringified input, but instead on something like `[<Struct>] type WarningDescription = Int32 of code:int | String of string | Ident of string` ?

Reply:
Let me try that. (I would probably use the original type `ParsedHashDirectiveArgument` then.)

Reply:
Why not wrap the CLI case into `(WarningDescription.String s)` struct wrapper?
It still holds a reference to the same string, its just wrapping it to tell the 3 cases apart.

Reply:
Ok, I did that, does not look that bad.